modules/plugin/mongodb/src/main/java/org/geotools/data/mongodb/AbstractFilterToMongo.java

```diff
@@ -527,14 +559,35 @@ public Object visit(DWithin filter, Object extraData) {
         Object e1 = filter.getExpression1().accept(this, Geometry.class);
 
         Geometry geometry = filter.getExpression2().evaluate(null, Geometry.class);
-        double distanceInMeters = DistanceBufferUtil.getDistanceInMeters(filter);
 
+        CoordinateReferenceSystem coordinateReferenceSystem =
+                featureType.getCoordinateReferenceSystem();
+        Unit<?> unit = coordinateReferenceSystem.getCoordinateSystem().getAxis(0).getUnit();
+        double distanceInMeters = DistanceBufferUtil.getDistanceInMeters(filter);
+        if (!METRIC_UNITS.contains(unit.toString())) {
+            MathTransform transform = null;
+            try {
+                // Transform geometry to CRS that supports meters
+                transform = CRS.findMathTransform(coordinateReferenceSystem, EPSG_3857);
```

Comment:
If the reference geometry is a point, I would leave the $near operator, it's likely to use proper ellipsoidic distances.

For all other geometries, EPSG:3857 is a poor choice for the task at hand, the deformation factor for lengths diverges towards the poles, and it's already around 2 times at 60° north/south.

A good projection for local geometries with equidistancy properties (in a small area at least) is the Azimuthal Equidistant, and there is an auto-code to re-center it. Suggestion:

* Determine the centroid of the geometry
* Create an azimuthal equidistant projection centered on it using the "AUTO:97003,9001,{lon},{lat}" projection, e.g., ``CRS.decode("AUTO:97003,9001,-110,20")``
* Then apply the same algorithm as you are now.

Mind, this will still just work for "small" areas , e.g., less than 100km from the center. For bigger geometries you'll probably have to split the reference geometry into multiple smaller bits, buffer each one within its own local projection, reproject back to native, then union them into a reference geometry.

Reply:
Refactored the code as proposed to make use of azimuthal equidistant projection. However one doubt I have is splitting the "big" geometry into smaller parts and processing them separately before merging them again. The implementation sounds not trivial and as I understand can affect the performance negatively. Checking the buffer applied to 2 linestrings, one along the shore of Norway and another from south pole to north I did not encountered any noticeable inaccuracies (visible to my eye), so I wonder if such if precision offered by splitting the geometry is required.

Reply:
You can start with the current implementation, it's better than nothing. If the distance happens to be greater than 100km maybe just log (at fine level) that the accuracy is likely not good enough.

Reply:
Added suggested logging in logAdditionalInformation method.